libpromises/matching.c

```diff
@@ -37,47 +37,45 @@
 
 
 /* Pure, non-thread-safe */
-static char *FirstBackReference(pcre *rx, const char *teststring)
+static char *FirstBackReference(Regex *regex, const char *teststring)
 {
     static char backreference[CF_BUFSIZE]; /* GLOBAL_R, no initialization needed */
-
-    int ovector[OVECCOUNT], i, rc;
-
     memset(backreference, 0, CF_BUFSIZE);
 
-    if ((rc = pcre_exec(rx, NULL, teststring, strlen(teststring), 0, 0, ovector, OVECCOUNT)) >= 0)
+    pcre2_match_data *match_data = pcre2_match_data_create_from_pattern(regex, NULL);
+    int result = pcre2_match(regex, (PCRE2_SPTR) teststring, PCRE2_ZERO_TERMINATED,
+                             0, 0, match_data, NULL);
+    /* pcre2_match() returns the highest capture group number + 1, i.e. 1 means
+     * a match with 0 capture groups. 0 means the vector of offsets is small,
+     * negative numbers are errors (incl. no match). */
+    if (result > 0)
     {
-        for (i = 1; i < rc; i++)        /* make backref vars $(1),$(2) etc */
+        size_t *ovector = pcre2_get_ovector_pointer(match_data);
+        /* ovector[0] and ovector[1] are for the start and end of the whole
+         * match, the capture groups follow in [1] and [2], etc. */
```

Comment:
did you mean [2] and [3]? since ovector[1] is the end of the whole match?

Reply:
https://github.com/cfengine/core/pull/5399